### PR TITLE
[DBEx 60966] add logging to uploaded evidence retrieval

### PR DESCRIPTION
**Pull request by SamStuckey**
## Summary

- Update and refactor for more robust and universal useability
    - `ThirdPartyTransactionLogging` (controller) concern into a universally usable object (lib).
    - add return of appropriate value. Previously the overwritten methods return value was lost
    - add additional logging of 'passed args'
    - current_user is now try(:current_user) to avoid a no method error outside of controller contexts.
    - change `breadcrumbs` log to `calling_method` and adjust logic to name the original objects method (the method we are logging around) rather than naming the logging method itself.
 - Expand Spec coverage around logging object for new return value logic.
 - Refactor `EVSS::DocumentUpload` for readability and loggability.
 - Add logging to key actions in the `EVSS::DocumentUpload` worker
 - remove deprecated and redundant `Datadog::Tracing.trace` usage.  
     - this pattern breaks our paradigm of using `Rails.logger` which is more universal
     - this pattern is redundant considering our Datadog logging config ([see datadog config](https://github.com/department-of-veterans-affairs/vets-api/blob/216ff7186c42a0db7cd7cc398d6c8529c3d178c3/config/initializers/datadog.rb#L19)
  - Minor refactor of previous 3PI logging POC in `UploadSupportingEvidencesController` for semantic correctness (upstream / downstream was used wrong).
 

## Related issue(s)
- [Ticket](https://app.zenhub.com/workspaces/disability-benefits-experience-team-carbs-6470c8bfffee9809b2634a52/issues/gh/department-of-veterans-affairs/va.gov-team/60966)
- [Epic](https://app.zenhub.com/workspaces/disability-benefits-experience-team-carbs-6470c8bfffee9809b2634a52/issues/gh/department-of-veterans-affairs/va.gov-team/60952)

## Testing done

- There is no measurable change beyond the creation of new logs.  Existing tests for the affected objects are sufficient to ensure code is working.
- Spec was added for inclusion of logging in new object type (worker)
- Spec was added for updated return value (true return from original method)

## What areas of the site does it impact?
- This will impact our form 526 submission flow by creating more detailed logs around supporting evidence retrieval prior to 526 form submission (to EVSS)

## Acceptance criteria

This feature is considered working and complete if
- the 526 form submission flow is not broken
- the 526 form submission flow now includes logging around supporting evidence retrieval from S3

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [x]  Documentation has been updated ([small adjustments to existing logging investigation](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/disability/526ez/engineering_research/error_handling_and_submission_failures/01_overview_of_error_and_logging.md) and [implementation documentation](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/disability/526ez/implementation/reusable_logging_with_third_party_transaction_logger.md)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [x]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected (N/A)
- [x]  I added a screenshot of the developed feature (N/A)

### CODE REVIEW

File: lib/logging/third_party_transaction.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module Logging
+  module ThirdPartyTransaction
```
Comment (SamStuckey) [inline lib/logging/third_party_transaction.rb:4]: This file was moved from `ThirdPartyTransactionLogging`, so unfortunately some git history was lost. Despite using `git mv`, the indentation of the new name spacing breaks git's ability to recognize this automatically as the same file.  There is probalby a way to do it manually, but I'm not sure it's worth the time.

Notable changes to this file are
1. location and namespace
2. addition of appropriate return value.  Previously the overwritten methods return value was lost
3. additional logging of 'passed args', which will probably be semi-useless stuff like <MyObject #23423498> most of the time, but imho is still nice to see.  This logging is added to both begin / complete methods

File: spec/lib/logging/third_party_transaction_spec.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Logging::ThirdPartyTransaction do
```
Comment (SamStuckey) [inline spec/lib/logging/third_party_transaction_spec.rb:5]: This file also lost some git history when moving.  Notable changes are:

1. location and namespace
2. refactoring from a single set of specs to a shared example group
3. addition of return value verification spec

File: spec/lib/shrine/plugins/validate_virus_free_spec.rb [hunk 3]
```diff
@@ -35,7 +35,7 @@ def errors
       context 'while in development' do
         it 'logs an error message if clamd is not running' do
           expect(Rails.env).to receive(:development?).and_return(true)
-          expect(Shrine.logger).to receive(:error).with(/PLEASE START CLAMD/)
+          expect(Rails.logger).to receive(:error).with(/PLEASE START CLAMD/)
```
Comment (SamStuckey) [inline spec/lib/shrine/plugins/validate_virus_free_spec.rb:38]: Shrine.logger in a Rails application simply delegates to `Rails.logger`, so we are changing it for simplicity and homogeneity.

File: lib/shrine/plugins/validate_virus_free.rb [hunk 4]
```diff
@@ -19,7 +19,7 @@ def validate_virus_free(message: nil)
 
         def add_error_msg(message)
           if Rails.env.development? && message.match(/nodename nor servname provided/)
-            Shrine.logger.error('VIRUS SCANNING IS OFF. PLEASE START CLAMD')
+            Rails.logger.error('VIRUS SCANNING IS OFF. PLEASE START CLAMD')
```
Comment (SamStuckey) [inline lib/shrine/plugins/validate_virus_free.rb:22]: Shrine.logger in a Rails application simply delegates to `Rails.logger`, so we are changing it for simplicity and homogeneity.

**Comment by SamStuckey** [comment]
Note that this PR as not as big as the linter thinks it is.  Two files were moved and namespaces, and the resulting indentation caused git to think I've deleted and recreated files.

**Comment by NB28VT** [comment]
This raises a worthwhile discussion around how we want to document changes. Ordinarily in something like this as a reviewer I'd want each change to be in its own commit message instead of summarized in the pull request to make it easier to track, but I'm not sure what the guidelines are on squashing and there are merits to having more space to document things in the PR. Maybe something we as a team settle on at some point?

File: app/workers/evss/document_upload.rb [hunk 5]
```diff
@@ -14,26 +30,53 @@ class EVSS::DocumentUpload
   end
 
   def perform(auth_headers, user_uuid, document_hash)
-    document, client, uploader, file_body = nil
```
Comment (NB28VT) [inline app/workers/evss/document_upload.rb:17]: (Blocking pending peace of mind answer) was the behavior you're (thankfully and successfully) refactoring here already tested and if so is it tested robustly? Just flagging this appears to be a relatively critical worker and it might be easy to miss something when doing this kind of refactor, as the original code is pretty hard to read.

Noting something like this would probably need to be QA'd/considered in a script for Test Rail as well but you may have already been anticipating that.

File: app/workers/evss/document_upload.rb [hunk 6]
```diff
@@ -14,26 +30,53 @@ class EVSS::DocumentUpload
...
-    end
+    @auth_headers = auth_headers
+    @user_uuid = user_uuid
+    @document_hash = document_hash
+
+    validate_document!
+    pull_file_from_cloud!
+    perform_document_upload_to_evss
+    clean_up!
+  end
```
Comment (NB28VT) [inline app/workers/evss/document_upload.rb:41]: Muchhhhhh better, big fan of this kind of pattern

**Review by sethdarragile6 — COMMENTED** [review]
lgtm!

File: lib/logging/third_party_transaction.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+          passed_args: args.to_s
+        }.merge(additional_logs)
+
+        Rails.logger.info(log)
+      rescue => e
+        Rails.logger.error(e)
+      end
+
+      def log_3pi_complete(method_name, additional_logs = {}, args = [])
+        now = Time.current
```
Comment (NB28VT) [inline lib/logging/third_party_transaction.rb:56]: (Blocking pending peace of mind answer) I'd double check the methods you are using for `Time` objects are consistent with other time-based logging elsewhere because of potential confusion related to timezones.

My understanding is that what are you using here, `Time.current`, which retrieves the time in the application's configured timezone, is the best method to use universally.

However, if you look at both our [application config](https://github.com/department-of-veterans-affairs/vets-api/blob/master/config/application.rb) and [production env config](https://github.com/department-of-veterans-affairs/vets-api/blob/master/config/environments/production.rb) it doesn't appear we configure the time zone for the application, and in that case I think it defaults to the time zone of whatever system the application is running on.

That may be ok but it's just something I thought was worth calling out in case you missed it.

File: spec/lib/logging/third_party_transaction_spec.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+      it 'returns the original methods return value' do
+        expect(test_object.method_to_wrap).to eq 'return value'
+      end
+    end
+
+    context 'when something goes wrong' do
+      it 'fails quietly and logs the problem' do
+        # Time is used in the logging methods
+        allow(Time).to receive(:current).and_raise(StandardError)
```
Comment (NB28VT) [inline spec/lib/logging/third_party_transaction_spec.rb:30]: (Blocking pending answer) Probably unlikely this will change but I'm not sure about this test picking one line of functionality within the method and mocking it failing. Is there an alternative approach/refactor that doesn't require this unit test knowing about a specific implementation detail of a method it is testing?

File: spec/lib/logging/third_party_transaction_spec.rb [hunk 9]
```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Logging::ThirdPartyTransaction do
+  # to use shared examples
```
Comment (NB28VT) [inline spec/lib/logging/third_party_transaction_spec.rb:6]: (Non blocking) Good call documenting this, might consider more specific language like "To use the shared examples below for the object you are testing" as here it sounds a little like you're documenting RSpec's `shared_examples_for` keyword

File: spec/lib/logging/third_party_transaction_spec.rb [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  # 1. define variable test_object of desired type using a `let` block
+  # 2. test_object must define an instance method `method_to_wrap`
+  # 3. test_object should extend Logging::ThirdPartyTransaction::MethodWrapper
+  # 4. test_object definition should end with a call to .new, ensuring an object
+  #    is returned to your variable, rather than the class definition block,
+  #    as our linter will reject this.
+  shared_examples_for 'a third party transaction logger' do
```
Comment (NB28VT) [inline spec/lib/logging/third_party_transaction_spec.rb:13]: (Non blocking) I forgot about shared examples this is handy. For only a handful of cases this feels like a little overkill  IMO since it requires the reader to understand the pattern you are using here but presumably if you want to test how this module works with other objects there will be more cases here and DRYing it all up makes sense

File: spec/rails_helper.rb [hunk 11]
```diff
@@ -32,6 +32,7 @@
 require 'super_diff/rspec-rails'
 require 'super_diff/active_support'
 require './spec/support/default_configuration_helper'
+require 'logging/third_party_transaction'
```
Comment (NB28VT) [inline spec/rails_helper.rb:35]: (Non blocking) Nitpicky but feel like I've been called out in the past for requiring modules for specific types of tests in global configuration files like `rails_helper`, as too many of these may slow down the test suite overall. I might consider requiring these somewhere else that is less global for the rest of the test suite

**Review by NB28VT — CHANGES_REQUESTED** [review]
Most of the "changes" I marked as "blocking" were just questions/callouts of things I noticed, so happy to approve after you've taken a look if you have a response that assuages any worries. The rest are probably more style and best practice preferences, feel free to ignore if you disagree

File: app/workers/evss/document_upload.rb (hunk 5 above)
Comment (SamStuckey) [reply app/workers/evss/document_upload.rb:17]: So yes this worker has a test suite : ).  This specific line is to make these variables available outside of the Datadog::Tracing blocks.  It's no longer necessary (and an anti-pattern besides) because of the loader methods I added below. 

But to address your core question, yes this worker is tested!

File: lib/logging/third_party_transaction.rb [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+    # these will be included after instance instantiation, making them available
+    # to the instance and retaining their scope.
+    module ScopedInstanceMethods
+      def log_3pi_begin(method_name, additional_logs = {}, args = [])
+        @start_time = Time.current
+
+        log = {
+          process_id: Process.pid,
+          user_uuid: try(:current_user).try(:account_uuid),
```
[the remark this answers is not in the document]
Comment (SamStuckey) [reply lib/logging/third_party_transaction.rb:43]: 1. Yes, I agree.  It would be cool to have this automagically scope itself to it's context, or exclude nil values from the log.  atm, i think this is fine, but yeah great note for a future refactor. 
2. `account_uuid` is  not necessarily universal, given (as you have seen) the weirdness or our User model.  I belive it may be dependent on how a user logs in.  I think another good future refactor could be changing this to `user_uuid: get_user_id` where `get_user_id`  tries a few values to return the salient user id 🤔 but... that's also a bit of weird indirection?  so I'd want to look into it before jumpting

tl;dr - good notes for future refactors imho

File: lib/logging/third_party_transaction.rb (hunk 7 above)
Comment (SamStuckey) [reply lib/logging/third_party_transaction.rb:56]: [PENDING ACTION] good idea, will do

File: spec/lib/logging/third_party_transaction_spec.rb (hunk 9 above)
Comment (SamStuckey) [reply spec/lib/logging/third_party_transaction_spec.rb:6]: [PENDING ACTION] will clean up the language

File: spec/lib/logging/third_party_transaction_spec.rb (hunk 10 above)
Comment (SamStuckey) [reply spec/lib/logging/third_party_transaction_spec.rb:13]: Yeah, shared examples are arguably an anti-patern, depending on who you ask, given the extra layer of abstraction and Ruby's stated mission of being readable.  But, in this case, it feels right?  ¯\_(ツ)_/¯

File: spec/lib/logging/third_party_transaction_spec.rb (hunk 8 above)
Comment (SamStuckey) [reply spec/lib/logging/third_party_transaction_spec.rb:30]: [PENDING ACTION\] So the point of the test is to force a failure to ensure quite failure, so as far as the stated purpose, i think this checks the box, and i think it's an important box to check.

That being said, yes I also kinda hate this.  I spent about an hour trying to come up with a better way and couldn't...   Def open to suggestions, I can also take another pass at it. 🤔

File: lib/logging/third_party_transaction.rb (hunk 7 above)
Comment (SamStuckey) [reply lib/logging/third_party_transaction.rb:56]: Confirmed that this same format is used in at least a high number of other cases.  here is an example from our [Webhooks::SchedulerJob](https://github.com/department-of-veterans-affairs/vets-api/blob/615c41468e1bbdb247664683aed4a5b82691ac74/app/workers/webhooks/scheduler_job.rb#L11).  I checked to see what `processing_time` was (you can see it set in the retry [a few lines down here](https://github.com/department-of-veterans-affairs/vets-api/blob/615c41468e1bbdb247664683aed4a5b82691ac74/app/workers/webhooks/scheduler_job.rb#L23)

Checking the class of `5.minutes.from_now` against `Time.current` reveals a match.  For sanity, here is what that output will look like in a log
`Fri, 28 Jul 2023 19:18:32.825530000 UTC +00:00`
I *cannot* confirm this is our only time format being logged, but at least we have a precedent, which feels good IMHO.

Screen Shot 2023-07-28 at 3 14 58 PM

File: spec/lib/logging/third_party_transaction_spec.rb (hunk 9 above)
Comment (SamStuckey) [reply spec/lib/logging/third_party_transaction_spec.rb:6]: done

File: spec/lib/logging/third_party_transaction_spec.rb (hunk 8 above)
Comment (SamStuckey) [reply spec/lib/logging/third_party_transaction_spec.rb:30]: On further thought, I'm not sure there is a better way to do this.  In order to test this functionality we need an error to be raised from *within* the tested method.  I really can't think of any other way to do this, but like I said I'm open to suggestions ¯\_(  ͠◔__  ͡◔)_/¯

File: spec/rails_helper.rb (hunk 11 above)
Comment (SamStuckey) [reply spec/rails_helper.rb:35]: done

File: spec/lib/logging/third_party_transaction_spec.rb (hunk 9 above)
Comment (NB28VT) [reply spec/lib/logging/third_party_transaction_spec.rb]: Oh another trick we had at my last gig, in responses like this replying "Done in XXXX" and linking the commit with the change. Again not being nitpicky 🤣 I just thought it was a good idea whoever introduced it there lol

**Review by NB28VT — APPROVED** [review]
LGTM!
